cli/build.go

```diff
@@ -434,40 +436,51 @@ func (b *BuildCmd) BuildInsideContainer(ctx context.Context, target *types.Build
 
 	log.Infof(ctx, "Building from %s in container: %s", containerWorkDir, buildContainer.Id)
 
-	// Local path to binary that we want to inject
-	// TODO: this only supports Linux containers
-	localYbPath := ""
-	devMode := false
-
 	// If this is development mode, use the YB binary currently running
 	// We can't do this by default because this only works if the host is
 	// Linux so we might as well behave the same on all platforms
 	// If not development mode, download the binary from the distribution channel
-	if b.Version == "DEVELOPMENT" {
-		if runtime.GOOS == "linux" {
-			// TODO: If we support building inside a container that's not Linux we will want to do something different
-			log.Infof(ctx, "Development version in use, will upload development binary to the container")
-			devMode = true
-		}
+	devMode := false
+	if b.Version == "DEVELOPMENT" && runtime.GOOS == "linux" {
+		// TODO: If we support building inside a container that's not Linux we will want to do something different
+		log.Infof(ctx, "Development version in use, will upload development binary to the container")
+		devMode = true
 	}
 
+	var localYB io.ReadCloser
+	var localYBSize int64
 	if devMode {
-		if p, err := os.Executable(); err != nil {
-			return fmt.Errorf("Can't determine local path to YB: %v", err)
-		} else {
-			localYbPath = p
+		p, err := os.Executable()
+		if err != nil {
+			return err
+		}
+		f, err := os.Open(p)
+		if err != nil {
+			return err
 		}
+		info, err := f.Stat()
+		if err != nil {
+			f.Close()
+			return err
+		}
+		localYB = f
+		localYBSize = info.Size()
 	} else {
-		if p, err := DownloadYB(ctx); err != nil {
-			return fmt.Errorf("Couldn't download YB: %v", err)
-		} else {
-			localYbPath = p
+		var err error
+		localYB, localYBSize, err = DownloadYB(ctx)
+		if err != nil {
+			return err
 		}
 	}
 
 	// Upload and update CLI
-	log.Infof(ctx, "Uploading YB from %s to /yb", localYbPath)
-	err = narwhal.UploadFile(ctx, dockerClient, buildContainer.Id, "/yb", localYbPath)
+	log.Infof(ctx, "Uploading YB to /yb")
+	err = narwhal.Upload(ctx, dockerClient, buildContainer.Id, "/yb", localYB, &tar.Header{
+		Mode:     0777,
+		Size:     localYBSize,
+		Typeflag: tar.TypeReg,
+	})
+	localYB.Close() // just a reader, Close can be ignored.
```

Comment:
Would this be more idiomatic in a `defer` near declaration or assignment?

Reply:
I was hesitant on this one because the containing function is large, but I solved by extracting this to a function.